File: c1_arrays_and_strings/cs/IsUnique.cs
// 1.1 implement an algorithm to determine if a string has all unique
// characters.
using System;
using System.Linq;
using System.Collections.Generic;

namespace IsUniqueApp {
  class IsUnique {

    private string DetermineIfCharsUnique(string givenString) {
      string retval = "";
      string[] uniqueChars = {};
      
      for (int i = 0; i < givenString.Length; i++) {
        if (Array.IndexOf(uniqueChars, Char.ToString(givenString[i])) == -1) {
          Array.Resize(ref uniqueChars, uniqueChars.Length + 1);
          uniqueChars[uniqueChars.Length - 1] = Char.ToString(givenString[i]);
        }
      }

      if (givenString.Length == uniqueChars.Length) {
        retval = "The string has only unique chars.";
      } else {
        retval = "This string has repeating characters.";
      }

      return retval;

    }

    public static void Main(string[] args) {

      IsUnique isUnique = new IsUnique();

      Console.WriteLine("Give me a string and I will determine if it is unique or not!");

      string givenString = Console.ReadLine();

      string result = isUnique.DetermineIfCharsUnique(givenString);

      Console.WriteLine(result);

    }
  }

}

File: c1_arrays_and_strings/ts/is-unique.ts
// 1.1 implement an algorithm to determine if a string has all unique
// characters.

const isUnique = (givenStr: string) => {
  // will contain a map of each character in the string to the number of times it shows up in the string
  const freqHash = {};

  for (let i = 0; i < givenStr.length; i++) {
    const char = givenStr[i];
    // check if a key of this value exists
    if (freqHash[char] === undefined) {
      freqHash[char] = 0;
    } else freqHash[char]++;
  }

  // iterate over entire hash map and if every value is 1 - return true, else false
  for (const [key, value] of Object.entries(freqHash)) {
    if (value !== 1) return false;
  }
  return true;
};

// test cases
